cmd/crank/xpkg/extract.go

```diff
@@ -0,0 +1,218 @@
+	errOpenPackageStream       = "failed to open package stream file"
+	errCreateOutputFile        = "failed to create output file"
+	errCreateGzipWriter        = "failed to create gzip writer"
+	errExtractPackageContents  = "failed to extract package contents"
+)
+
+const (
+	layerAnnotation     = "io.crossplane.xpkg"
+	baseAnnotationValue = "base"
+	cacheContentExt     = ".gz"
+)
+
+// fetchFn fetches a package from a source.
+type fetchFn func(context.Context, name.Reference) (v1.Image, error)
+
+// registryFetch fetches a package from the registry.
+func registryFetch(ctx context.Context, r name.Reference) (v1.Image, error) {
+	return remote.Image(r, remote.WithContext(ctx))
+}
+
+// daemonFetch fetches a package from the Docker daemon.
+func daemonFetch(ctx context.Context, r name.Reference) (v1.Image, error) {
+	return daemon.Image(r, daemon.WithContext(ctx))
+}
+
+func xpkgFetch(path string) fetchFn {
+	return func(_ context.Context, _ name.Reference) (v1.Image, error) {
+		return tarball.ImageFromPath(filepath.Clean(path), nil)
+	}
+}
+
+// AfterApply constructs and binds context to any subcommands
+// that have Run() methods that receive it.
+func (c *extractCmd) AfterApply() error {
+	c.fs = afero.NewOsFs()
+	c.fetch = registryFetch
+	if c.FromDaemon {
+		c.fetch = daemonFetch
+	}
+	if c.FromXpkg {
+		// If package is not defined, attempt to find single package in current
+		// directory.
+		if c.Package == "" {
+			wd, err := os.Getwd()
+			if err != nil {
+				return errors.Wrap(err, errGetwd)
+			}
+			path, err := xpkg.FindXpkgInDir(c.fs, wd)
+			if err != nil {
+				return errors.Wrap(err, errFindPackageinWd)
+			}
+			c.Package = path
+		}
+		c.fetch = xpkgFetch(c.Package)
+	}
+	if !c.FromXpkg {
+		if c.Package == "" {
+			return errors.New(errMustProvideTag)
+		}
+		upCtx, err := upbound.NewFromFlags(c.Flags)
+		if err != nil {
+			return err
+		}
+
+		name, err := name.ParseReference(c.Package, name.WithDefaultRegistry(upCtx.RegistryEndpoint.Hostname()))
+		if err != nil {
+			return errors.Wrap(err, errInvalidTag)
+		}
+		c.name = name
+	}
+	return nil
+}
+
+// extractCmd extracts package contents into a Crossplane cache compatible
+// format.
+type extractCmd struct {
+	fs    afero.Fs
+	name  name.Reference
+	fetch fetchFn
+
+	Package    string `arg:""                                                                                                                                                     help:"Name of the package to extract. Must be a valid OCI image tag or a path if using --from-xpkg." optional:""`
+	FromDaemon bool   `help:"Indicates that the image should be fetched from the Docker daemon."                                                                                  xor:"xp-extract-from"`
+	FromXpkg   bool   `help:"Indicates that the image should be fetched from a local xpkg. If package is not specified and only one exists in current directory it will be used." xor:"xp-extract-from"`
```

Comment:
i don't think the args referenced by these `xor` fields still exist, can they be removed?